src/utils.cpp

```diff
@@ -738,6 +738,45 @@ size_t utils::wcswidth_stfl(const std::wstring& str, size_t size) {
 	return width - reduce_count;
 }
 
+std::string utils::substr_with_width(
+		const std::string& str,
+		const size_t max_width)
+{
+	if (str.empty() || !max_width) {
+		return std::string("");
+	}
+
+	const std::wstring wstr = utils::str2wstr(str);
+	size_t begin = 0;
+	size_t count = std::ceil(wstr.length() / 2.0);
+	size_t checked_width = 0;
+	size_t part_width = 0;
+	int iter_limit = 20;
+
+	while (iter_limit > 0 && count) {
```

Comment:
Instead of this algorithm, why not iterate through `wstr` accumulating results of `wcwidth` and exit when the accumulator reaches `max_width`?

Reply:
Oh, and the loop should also skip over `<` and `>`. In your implementation, `utils::strwidth_stfl` already does that, but in the algorithm I propose, it'll have to be done again.